File: QEasyWidgets/Components/List.py
from typing import Optional, overload
from PySide6.QtGui import *
from PySide6.QtCore import *
from PySide6.QtWidgets import *

from ..Common.StyleSheet import *
from ..Common.QFunctions import *
from .Menu import MenuBase

##############################################################################################################################

class ListBase(QListView):
    """
    Base class for list components
    """
    _contextMenu = None

    currentItemChanged = Signal(QStandardItem, QStandardItem)
    itemClicked = Signal(QStandardItem)

    def __init__(self, parent: Optional[QWidget] = None):
        super().__init__(parent)

        self.StandardItemModel = QStandardItemModel(self)
        self.setModel(self.StandardItemModel)

        self.clicked.connect(self.onItemClicked)

        setFont(self, 12)

        StyleSheetBase.List.apply(self)

    def model(self) -> QStandardItemModel:
        return self.StandardItemModel

    def _toStandardItem(self, item):
        if isinstance(item, QListWidgetItem):
            item = QStandardItem(item.text())
        if isinstance(item, str):
            item = QStandardItem(item)
        return item

    def row(self, item):
        item = self._toStandardItem(item)
        index = self.model().indexFromItem(item)
        return index.row()

    def count(self):
        return self.model().rowCount()

    def item(self, row):
        return self.model().item(row)

    def addItem(self, item):
        item = self._toStandardItem(item)
        self.model().appendRow(item)

    def takeItem(self, row):
        return self.model().takeRow(row)[0] # takeRow returns a list

    def currentItem(self):
        indexes = self.selectedIndexes()
        if indexes:
            return self.model().itemFromIndex(indexes[0])
        return None

    def setCurrentItem(self, item):
        item = self._toStandardItem(item)
        index = self.model().indexFromItem(item)
        self.setCurrentIndex(index)

    def clear(self):
        self.model().clear()

    def click(self, item):
        item = self._toStandardItem(item)
        index = self.model().indexFromItem(item)
        self.setCurrentIndex(index)
        self.clicked.emit(index)

    @Slot('QModelIndex')
    def onItemClicked(self, index: Union[QModelIndex, QPersistentModelIndex]):
        item = self.model().itemFromIndex(index)
        self.itemClicked.emit(item)

    def currentChanged(self, current: Union[QModelIndex, QPersistentModelIndex], previous: Union[QModelIndex, QPersistentModelIndex]):
        super().currentChanged(current, previous)
        currentItem, previousItem = self.model().itemFromIndex(current), self.model().itemFromIndex(previous)
        self.currentItemChanged.emit(currentItem, previousItem)

    @property
    def contextMenu(self):
        if self._contextMenu is None:
            self._contextMenu = MenuBase(self)
        return self._contextMenu

    @contextMenu.setter
    def contextMenu(self, menu: MenuBase):
        ''''''
        self._contextMenu = menu

    def setContextMenu(self, actions: dict)-> None:
        self.setContextMenuPolicy(Qt.CustomContextMenu) if self.contextMenuPolicy() != Qt.CustomContextMenu else None
        self.customContextMenuRequested.connect(
            lambda position: (
                self.contextMenu.clear(),
                showContextMenu(self, self.contextMenu, actions, self.mapToGlobal(position))
            )
        )

    def setBorderless(self, borderless: bool) -> None:
        self.setProperty("isBorderless", borderless)

    def clearDefaultStyleSheet(self) -> None:
        StyleSheetBase.List.deregistrate(self)

##############################################################################################################################

File: c++/QEasyWidgets/Components/List.cpp
#include "List.h"

#include <QFont>


namespace QEW {

ListBase::ListBase(QWidget *parent)
    : QListView(parent)
    , m_model(nullptr)
{
    init();
}

void ListBase::init()
{
    m_model = new QStandardItemModel(this);
    setModel(m_model);

    connect(this, &QListView::clicked, this, &ListBase::onItemClicked);

    QFont font = this->font();
    font.setPointSize(12);
    setFont(font);

    // TODO: Apply stylesheet
    // StyleSheetBase::apply(this, StyleSheetBase::List);
}

QStandardItemModel *ListBase::model() const
{
    return m_model;
}

int ListBase::row(const QStandardItem *item) const
{
    if (!item) return -1;
    QModelIndex index = m_model->indexFromItem(item);
    return index.row();
}

void ListBase::insertRow(int row, QStandardItem *item)
{
    if (item) {
        m_model->insertRow(row, item);
    }
}

void ListBase::removeRow(int row)
{
    m_model->removeRow(row);
}

int ListBase::rowCount() const
{
    return m_model->rowCount();
}

void ListBase::setRowCount(int rows)
{
    m_model->setRowCount(rows);
}

QStandardItem *ListBase::item(int row) const
{
    return m_model->item(row);
}

void ListBase::setItem(int row, QStandardItem *item)
{
    if (item) {
        m_model->setItem(row, item);
    }
}

void ListBase::onItemClicked(const QModelIndex &index)
{
    if (index.isValid()) {
        QStandardItem *item = m_model->itemFromIndex(index);
        emit itemClicked(item);
    }
}

} // namespace QEW
